### PR TITLE
Added Tanh, Arctan, Softplus, Sigmoid to activation functions

### PR DESCRIPTION
**Pull request by Jashcraf**
Thanks to [this handy article](https://www.sciencedirect.com/topics/engineering/activation-function) I was able to code up some new activation functions with some additional flexibility (x,y offset, slope scaling). I was hoping to play with these using `dygdug` for coronagraph design sometime in the near future.

Tested against `scipy.optimize.approx_fprime`
image

Code to generate these results
```python
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import approx_fprime
plt.style.use('bmh')

from prysm.x.optym.activation import(
    Tanh,
    Arctan,
    Softplus,
    Sigmoid
)

x = np.linspace(-2,2,100)
a, x0, y0 = 1, 1, 1

tanh = Tanh(a, x0, y0)
atan = Arctan(a, x0, y0)
soft = Softplus(a, x0, y0)
sigm = Sigmoid(a, x0, y0)

tanh_f = tanh.forward(x)
atan_f = atan.forward(x)
soft_f = soft.forward(x)
sigm_f = sigm.forward(x)

tanh_g = tanh.backprop(x)
atan_g = atan.backprop(x)
soft_g = soft.backprop(x)
sigm_g = sigm.backprop(x)

tanh_fp = []
atan_fp = []
soft_fp = []
sigm_fp = []

for u in x:
    tanh_fp.append(approx_fprime(u, tanh.forward, 1e-8))
    atan_fp.append(approx_fprime(u, atan.forward, 1e-8))
    soft_fp.append(approx_fprime(u, soft.forward, 1e-8))
    sigm_fp.append(approx_fprime(u, sigm.forward, 1e-8))

plt.figure()
plt.subplot(221)
plt.title('Tanh')
plt.plot(x, tanh_f, label='forward')
plt.plot(x, tanh_g, label='backprop')
plt.plot(x, tanh_fp, label='apprx fprime', linestyle='dashed')
plt.legend()
plt.subplot(222)
plt.title('Arctan')
plt.plot(x, atan_f, label='forward')
plt.plot(x, atan_g, label='backprop')
plt.plot(x, atan_fp, label='apprx fprime', linestyle='dashed')
plt.legend()
plt.subplot(223)
plt.title('Softplus')
plt.plot(x, soft_f, label='forward')
plt.plot(x, soft_g, label='backprop')
plt.plot(x, soft_fp, label='apprx fprime', linestyle='dashed')
plt.legend()
plt.subplot(224)
plt.title('Sigmoid')
plt.plot(x, sigm_f, label='forward')
plt.plot(x, sigm_g, label='backprop')
plt.plot(x, sigm_fp, label='apprx fprime', linestyle='dashed')
plt.legend()
plt.show()
```

### CODE REVIEW

File: prysm/x/optym/activation.py [hunk 1]
```diff
@@ -206,3 +206,110 @@ def discretize(self, x):
         # take argmax along dim k, and take that from levels
         indices = np.argmax(encoded, axis=-1)
         return np.take(self.levels, indices)
+    
+
+class Tanh:
+    """Tanh activation function.
+
+    Uses the hyperbolic tangent function f(x) = tanh(x). Constructed to support additional
```
Comment (brandondube) [inline prysm/x/optym/activation.py:214]: For all of these functions, the docstring is of the form

```
"""Name to satisfy linter.

Uses the f(x) = the_f(x)
...
"""
```

I think the line starting with Uses is not value added?

I would do just

```
class Tanh:
"""Tanh(x).

def __init__(self, a=1, x0=0, y0=0): pass
```

File: prysm/x/optym/activation.py [hunk 2]
```diff
@@ -206,3 +206,110 @@ def discretize(self, x):
...
+        fx = self.forward(xbar) - self.y0 # have to subtract offset
+        return self.a*(1 - fx**2) 
+
+
+class Arctan:
+    """Arctan activation function.
+
+    Uses the inverse tangent function f(x) = arctan(x). Constructed to support additional
+    free parameters at initialization, namely:
+    - a : scale the slope of the function
```
Comment (brandondube) [inline prysm/x/optym/activation.py:242]: please document a, x0, y0 in `__init__`, in numpy docstring format

**Review by brandondube — COMMENTED** [review]
I left two comments.  Could you also add a tests file that verifies the backpropagation against finite differences (approx_fprime)?

**Comment by Jashcraf** [comment]
Made the docstring fixes, what directory would you like the tests in? I think I put the tests for x/pol in the x/ directory, but should the x/activation tests be in the /activation/ directory?

**Comment by brandondube** [comment]
tests can go in `x/optym/test_activation.py` for now; they eventually need to go in tests/, but you need to tell pytest to ignore them for now or they will ding the coverage score.  I had planned to put all x/ tests in the right spot when a module becomes not-x/

**Comment by Jashcraf** [comment]
Tests added!

**Pull request merged**